### PR TITLE
[POAE7-2638] Refactor CiderOperator to support new Join Design

### PR DESCRIPTION
**Pull request by chaojun-zhang**
### What changes were proposed in this pull request?

Add CiderPipelineOperator to support new join design , current just provide some interface will raise future PR for this.

Add basic BatchProcessor interfaces

### Why are the changes needed?

Support new Join design

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

not tested.
### Which label does this PR belong to?

cider, veloxIntegration

### CODE REVIEW

File: cider/include/cider/processor/BatchProcessor.h [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+  kFinished,
+};
+
+class BatchProcessor {
+ public:
+  ~BatchProcessor() = default;
+
+  BatchProcessorContextPtr getContext() const { return context_; }
+
+  virtual void processNextBatch(std::shared_ptr<CiderBatch> batch);
```
Comment (jikunshang) [inline cider/include/cider/processor/BatchProcessor.h:43]: For this set of API,  should we still use CiderBatch or use ArrowArray? cc@xieqi

File: cider-velox/src/CiderVeloxOptions.cpp [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#include "CiderVeloxOptions.h"
+
+DEFINE_bool(use_next_gen_compiler, false, "Use next generation compiler");\
```
Comment (jikunshang) [inline cider-velox/src/CiderVeloxOptions.cpp:24]: should be defined in cider module.

File: cider-velox/src/CiderPipelineOperator.h [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+#pragma once
+
+#include "CiderOperator.h"
+#include "CiderPipelineOperator.h"
+#include "cider/processor/BatchProcessor.h"
+
+namespace facebook::velox::plugin {
+
+class CiderPipelineOperator : public CiderOperator {
```
Comment (jikunshang) [inline cider-velox/src/CiderPipelineOperator.h:30]: i feel it should no longer extends `CiderOperator`

Comment (chaojun-zhang) [reply cider-velox/src/CiderPipelineOperator.h:30]: Yes, will remove once next gen compiler works well

File: cider-velox/src/CiderPipelineOperator.cpp [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+  return dataConvertor_->convertToRowVector(
+      *output_, *outputSchema_, operatorCtx_->pool());
+}
+
+CiderPipelineOperator::CiderPipelineOperator(
+    int32_t operatorId,
+    exec::DriverCtx* driverCtx,
+    const std::shared_ptr<const CiderPlanNode>& ciderPlanNode)
+    : CiderOperator(operatorId, driverCtx, ciderPlanNode) {
+  auto context = std::make_shared<cider::processor::BatchProcessorContext>();
```
Comment (harborn) [inline cider-velox/src/CiderPipelineOperator.cpp:94]:
```c++
auto context = std::make_shared<const cider::processor::BatchProcessorContext>();
```

File: cider-velox/src/CMakeLists.txt [hunk 5]
```diff
@@ -31,7 +31,9 @@ set(VELOX_PLUGIN_SOURCES
     ArrowConvertorUtils.cpp
     CiderVeloxPluginCtx.cpp
     CiderStatefulOperator.cpp
-    CiderStatelessOperator.cpp)
+    CiderStatelessOperator.cpp
+    CiderVeloxOptions.cpp
+    CiderPipelineOperator.cpp)
```
Comment (harborn) [inline cider-velox/src/CMakeLists.txt:36]: add target_link_libraries cider_processor, or compile failed with info 'undefined reference'

File: cider/include/cider/processor/BatchProcessor.h (hunk 1 above)
Comment (chaojun-zhang) [reply cider/include/cider/processor/BatchProcessor.h:43]: we discuss with @xieqi for the next pr

File: cider/include/cider/processor/BatchProcessor.h [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+#ifndef CIDER_BATCH_PROCESSOR_H
+#define CIDER_BATCH_PROCESSOR_H
+
+#include "BatchProcessorContext.h"
+#include "cider/CiderBatch.h"
+#include "substrait/plan.pb.h"
+
+namespace cider::processor {
+
+enum BatchProcessorState {
```
Comment (xieqi) [inline cider/include/cider/processor/BatchProcessor.h:31]: better enum class

File: cider-velox/src/CiderVeloxOptions.cpp (hunk 2 above)
Comment (chaojun-zhang) [reply cider-velox/src/CiderVeloxOptions.cpp]: REMOVED

File: cider/include/cider/processor/BatchProcessorContext.h [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+#ifndef CIDER_BATCH_PROCESSOR_CONTEXT_H
+#define CIDER_BATCH_PROCESSOR_CONTEXT_H
+
+#include <memory>
+
+namespace cider::processor {
+
+class BatchProcessorContext {};
+
+using BatchProcessorContextPtr = std::shared_ptr<const BatchProcessorContext>;
```
Comment (xieqi) [inline cider/include/cider/processor/BatchProcessorContext.h:31]: need add const? will never change BatchProcessorContext?

**Review by xieqi — APPROVED** [review]

File: cider-velox/src/CiderPipelineOperator.cpp [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+#include "velox/vector/arrow/Abi.h"
+#include "velox/vector/arrow/Bridge.h"
+
+namespace facebook::velox::plugin {
+
+bool CiderPipelineOperator::needsInput() const {
+  return !noMoreInput_ && !input_;
+}
+
+void CiderPipelineOperator::addInput(facebook::velox::RowVectorPtr input) {
```
Comment (xieqi) [inline cider-velox/src/CiderPipelineOperator.cpp:33]: no need facebook::velox::

File: cider/include/cider/processor/BatchProcessorContext.h (hunk 7 above)
Comment (chaojun-zhang) [reply cider/include/cider/processor/BatchProcessorContext.h]: only changed  in HashJoinChannel

File: cider/include/cider/processor/BatchProcessor.h (hunk 6 above)
Comment (chaojun-zhang) [reply cider/include/cider/processor/BatchProcessor.h]: updated

File: cider-velox/src/CiderPipelineOperator.cpp (hunk 8 above)
Comment (chaojun-zhang) [reply cider-velox/src/CiderPipelineOperator.cpp]: Fixed

Comment (xieqi) [reply cider-velox/src/CiderPipelineOperator.cpp]: still has some unnecessary Facebook::velox::

**Comment by chaojun-zhang** [comment]
@xieqi  @harborn please help to review

**Review by xieqi — APPROVED** [review]

**Review by harborn — APPROVED** [review]

**Pull request merged**